### PR TITLE
[WIP] Build improvements, for users and for Jenkins

### PR DESCRIPTION
**Pull request by larroy**
## Description ##
Provide an easy to use build.py script to aglutinate coplex build logic which works on multiple platforms and does a build process plus artifacts that can be easily used.

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by larroy** [comment]
@marcoabreu 
@mbaijal 
@KellenSunderland

**Comment by marcoabreu** [comment]
I would like to keep the parameters in the jenkinsfile. Please make it possible that they can be passed into your python script.
I'll make a detailed review tomorrow.

**Comment by larroy** [comment]
Why do you want to keep the parameters in the Jenkinsfile?  From what I have seen in the past from great test engineers it that they keep the config in jenkins to a minimum and directly call a script.  I don't like all kinds of flags and config in the Jenkins configuration, for starters because it's harder to reproduce when you build in the platform by yourself to fix issues. (As a developer I don't want to specify all these flags when building).

**Comment by larroy** [comment]
I actually was thinking of having a build.json file with all the flags that can be easily modified.

File: Jenkinsfile [hunk 1]
```diff
@@ -228,31 +228,10 @@ try {
...
-              xcopy mshadow\\mshadow pkg_vc14_cpu\\include\\mshadow /E /I /Y
-              xcopy nnvm\\include pkg_vc14_cpu\\nnvm\\include /E /I /Y
-              del /Q *.7z
-              7z.exe a vc14_cpu.7z pkg_vc14_cpu\\
-              '''
-            stash includes: 'vc14_cpu.7z', name: 'vc14_cpu'
-          }
+          init_git_win()
+          bat '''
+            python build.py
```
Comment (piiswrong) [inline Jenkinsfile]: doesn't look like a good idea to use python. Why not build.bat?

File: build.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        try:
+            os.mkdir(p)
+        except OSError as e:
+            if e.errno != 17:
+                raise
+
+
+
+def windows_build(args):
+    BUILD_BAT=r'''call "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\x86_amd64\vcvarsx86_amd64.bat"
```
Comment (piiswrong) [inline build.py]: This isn't generic enough to be a script under root

Comment (larroy) [reply build.py]: the point is to grow it so you can just call build.py on any platform, right now building is a mess.

**Pull request closed without merging**